pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -676,73 +678,93 @@ private void completeConnect(int clientProtoVersion, String clientVersion, boole
+                //    a `CommandConnected` response
+                // 2. an authentication refresh, in which case we need to refresh authenticationData and role
+                String newAuthRole;
+                try {
+                    newAuthRole = authState.getAuthRole();
+                } catch (AuthenticationException e) {
+                    return CompletableFuture.failedFuture(e);
+                }
+                AuthenticationDataSource newAuthDataSource = authState.getAuthDataSource();
+                if (useOriginalAuthState) {
+                    this.originalAuthData = newAuthDataSource;
+                } else {
+                    this.authRole = newAuthRole;
+                    this.authenticationData = newAuthDataSource;
+                }
 
-            if (state != State.Connected) {
-                // First time authentication is done
-                completeConnect(clientProtocolVersion, clientVersion, enableSubscriptionPatternEvaluation);
-            } else {
-                // If the connection was already ready, it means we're doing a refresh
-                if (!StringUtils.isEmpty(authRole)) {
-                    if (!authRole.equals(newAuthRole)) {
-                        log.warn("[{}] Principal cannot change during an authentication refresh expected={} got={}",
-                                remoteAddress, authRole, newAuthRole);
-                        ctx.close();
-                    } else {
-                        log.info("[{}] Refreshed authentication credentials for role {}", remoteAddress, authRole);
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Client successfully authenticated with {} role {} and originalPrincipal {}, "
+                                    + "using original auth state: {}",
+                            remoteAddress, authMethod, this.authRole, originalPrincipal, originalAuthState);
+                }
+                if (state != State.Connected) {
+                    // First time authentication is done
+                    completeConnect(clientProtocolVersion, clientVersion, enableSubscriptionPatternEvaluation);
+                } else {
+                    // If the connection was already ready, it means we're doing a refresh
+                    if (!StringUtils.isEmpty(authRole)) {
+                        if (!authRole.equals(newAuthRole)) {
+                            return CompletableFuture.failedFuture(new AuthenticationException(String.format(
+                                    "Principal cannot change during an authentication refresh expected=%s got=%s",
+                                    authRole, newAuthRole)));
+                        } else {
+                            log.info("[{}] Refreshed authentication credentials for role {}", remoteAddress, authRole);
+                        }
                     }
+                    state = State.Connected;
```

Comment:
And if we can reach here. The state should be Connected right? Why need to set to Connected again?

And we should avoid change the state in another thread. Here is executed by the callback thread for completing the `authFuture`.

Reply:
> And if we can reach here. The state should be Connected right? Why need to set to Connected again?

You are right.

> And we should avoid change the state in another thread. Here is executed by the callback thread for completing the authFuture.

Do you mean I should switch to Pulsar thread to change the state?

Reply:
> And we should avoid change the state in another thread. Here is executed by the callback thread for completing the authFuture.

This is safe, each connection has its own ServerCnx.java, don't warry.

Reply:
I would like to know why this is not a safe operation. Are you worried that some other handler will change this state?

> I think we should remove all of the async references in this PR and just focus on fixing the underlying problem with authentication data.

Good point out.